### PR TITLE
Remove all nulls

### PR DESCRIPTION
**Pull request by OlivierBlanvillain**
This PR refactors the entire code base to remove all usages for null as a values, and all null checks. When needed, data structures are updated to use `java.util.Optional`, but in a lot of cases null are removed without replacement.

The second commit re-implements `nextBranch` and `previousBranch` in functional style (best reviewed commit by commit).

We should probably branch off for v0.6 before merging this in case I missed something, but if possible I would like to avoid rebasing this for too long ^^'

### CODE REVIEW

**Comment by featurecat** [comment]
wow, thanks :). I'll try to finish up all of the stuff for 0.6 now.

**Comment by zsalch** [comment]
I'm checking.
However, there are still many PRs that do not merge. Will it be a bit early? Next, is these PR changes according to this rule, or are these PR changes during the merge?

**Comment by OlivierBlanvillain** [comment]
Do you have any PR you would really like to get in before 0.6? I can focus on reviewing them this week. The fear is that if we wait for the PR queue to be empty but continue working like we do now, we will never release. I think it would be fine to release 0.6, and 0.7 in a month or something...

> Next, is these PR changes according to this rule, or are these PR changes during the merge?

Either way, but I will try to keep the workload related to this PR on my shoulders. I'm happy to get everything merged first, then I will rebase this PR on top of whatever changed, or the other way around, we could merge this first and I will help fixing and updating other PRs like I did for the big one on auto formatting.

**Comment by zsalch** [comment]
Ok, let's finish it together and I will complete the review as soon as possible. I will not submit a new PR until then.

**Comment by zsalch** [comment]
I reviewed all the changes and there was basically no problem. There are a few minor questions, I will test it later.

File: src/main/java/featurecat/lizzie/gui/BoardRenderer.java [hunk 1]
```diff
@@ -498,13 +490,13 @@ private void drawMoveNumbers(Graphics2D g) {
...
         int here = Board.getIndex(i, j);
-        Stone stoneHere = branch == null ? board.getStones()[here] : branch.data.stones[here];
+        Stone stoneHere = branchOpt.map(b -> b.data.stones[here]).orElse(board.getStones()[here]);
 
         // don't write the move number if either: the move number is 0, or there will already be
         // playout information written
         if (moveNumberList[Board.getIndex(i, j)] > 0
-            && !(branch != null && Lizzie.frame.isMouseOver(i, j))) {
-          if (lastMove != null && i == lastMove[0] && j == lastMove[1])
+            && (branchOpt.isPresent() || !Lizzie.frame.isMouseOver(i, j))) {
```
[the remark this answers is not in the document]
Comment (OlivierBlanvillain) [reply src/main/java/featurecat/lizzie/gui/BoardRenderer.java]: Yes, I messed up the conversion, good catch!

**Comment by zsalch** [comment]
@OlivierBlanvillain 
If fix a few problems that have been discovered, the other looks good for me.
I have merged these changes to my source, and by using it, it seems that there are not too many problems.
I hope to merge this PR into Next earlier, so new commits and merges can be made.

**Comment by zsalch** [comment]
LGTM.

**Pull request merged**